ENH: add label, title etc. to gridmapdisplay

gamaanderson opened the pull request:
Several additions to gridmapdisplay, making it more similar to radardisplay. I'm not sure how some things here will be received, so while revising this keep in mind my last comment to issue  #294.

Also check pyart example `plot_grid_three_panel.py` for eventual bugs

gamaanderson commented on pyart/graph/gridmapdisplay.py:
```diff
@@ -62,9 +64,8 @@ def __init__(self, grid, debug=False):
-        self.proj = pyproj.Proj(proj='lcc', datum='NAD83',
-                                lat_0=lat0, lon_0=lon0,
-                                x_0=0.0, y_0=0.0)
+        self.proj = pyproj.Proj(proj='aeqd', datum='NAD83',
```
I has having "huge" error with lamber, so I changed the data projection to azimuthal equidistant

gamaanderson commented on pyart/graph/gridmapdisplay.py:
```diff
@@ -328,11 +653,65 @@ def _check_basemap(self):
+        basemap = Basemap(llcrnrlon=min_lon, llcrnrlat=min_lat,
+                          urcrnrlon=max_lon, urcrnrlat=max_lat,
+                          lat_0=lat_0, lon_0=lon_0,
+                          projection='tmerc', area_thresh=area_thresh,
```
I prefer mercator to miller as most maps (e.g google maps) use it and it also perverse format. `'tmerc'` work better than `'merc'`, setting the scale from origin and it would also set the axes origin if it were not for [this basemap bug](https://github.com/matplotlib/basemap/issues/192)

gamaanderson commented on pyart/graph/gridmapdisplay.py:
```diff
@@ -348,31 +727,144 @@ def _find_nearest_grid_indices(self, lon, lat):
+    #      get methods       #
+    ##########################
+
+    def get_basemap(self):
```
I will be using basemap in artview to convert from map-projection to data-projection, those this need to be public

jjhelmus replied:
Can ARTView access the basemap using attribute access using display.basemap or getattr(display, 'basemap') so that this can be a private function?  If not, that is fine also just want to check.

gamaanderson replied:
Yes, we can access the attribute, but as there is no distinction between public and private attributes someone may want in the future to change `basemap` (e.g to `base_map`). The use of a public function is kind of a safety.

gamaanderson commented:
about the travis error, it is because I'm not raising the "no basemap" error but rather running it: if basemap has no mandatory arguments why no run it for the user if he didn't do it him self?

jjhelmus commented:
I think that particular test can be modified, it seem reasonable to create a default basemap instance if one does not exist prior to a call to `plot_grid`.  Should the same behavior be followed in `plot_crosshairs`?  I vote yes.  If so then the `_check_basemap` can be removed completely.

jjhelmus commented on pyart/graph/gridmapdisplay.py:
```diff
@@ -107,38 +109,28 @@ def plot_basemap(self, lat_lines=None, lon_lines=None,
+                                   labels=[False, False, False, True])
+
+    def plot_grid(self, field, level=0, vmin=None, vmax=None, cmap='jet',
+                  mask_outside=True, title=None, title_flag=True,
```
I think the default for mask_outside should be `False` as in the **RadarDisplay** class.

gamaanderson replied:
the default is `True` in [ppi](https://github.com/ARM-DOE/pyart/blob/master/pyart/graph/radardisplay.py#L265) and [rhi](https://github.com/ARM-DOE/pyart/blob/master/pyart/graph/radardisplay.py#L365) !?! But if it is to be `False` I have no objection

jjhelmus replied:
We talked about changing it to False in all instances in #335 but no ticket was created for the task.  For the time being keep this as `True` to be consistent with the other plotting function (except plot_ray) which happened to be the first one I found.  Oops

jjhelmus replied:
Issue #351 created to track changing mask_outside to `False` in all instances.

gamaanderson commented:
I also vote yes for default `Basemap` to `plot_crosshairs`

jjhelmus commented:
Should the same projections be used to create the basemap instance be used to project the latitudes and longitude positions?  Also, if the projection name can be made a parameter to the `__init__` function or the `_make_basemap function` that would be great.

I do not think the **grid_lons** and **grid_lats** class attributes are needed, they can be calculated in the `plot_grid` and `_make_basemap` functions.  This required a bit of work, do not feel obligated to refactor these out if it does not interest you.  I think the same can be said for **self.prog** but is a bit more tricky.

gamaanderson commented:
I didn't understand your first question, could you clarify it.

passing the projection to `plot_basemap` and them to `_make_basemap` would be nice.

I agree with removing **grid_lons** and **grid_lats**.

Ultimately **self.proj** should be provided by the grid it self, while that does not happen we could let as it is.

scollis commented:
We have a ticket on having a proj class.. I will update this one soon.. 
and I hope to get to fixing the whole projection issue next week


On 8/6/15 3:00 PM, Anderson wrote:
>
> I didn't understand your first question, could you clarify it.
>
> passing the projection to |plot_basemap| and them to |_make_basemap| 
> would be nice.
>
> I agree with removing *grid_lons* and *grid_lats*.
>
> Ultimately *self.proj* should be provided by the grid it self, while 
> that does not happen we could let as it is.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ARM-DOE/pyart/pull/346#issuecomment-128487046>.
>

-- 
I ride for Parkinsons research http://www.events.org/sponsorship.aspx?id=51573

jjhelmus commented:
In `__init__` the aeqd projection is used to create self.proj and then in `_make_basemap` a tmerc projection is used.  Should these be the same?

gamaanderson commented:
@scollis, we are kind of going back-wards, we should first solve issue #285, then #294 and at last solve the graphing. Of course, I was the one start with the last, so I can't really complain about that.

scollis commented:
yep.. but I will make a general proj class that returns an instance to 
do projections..
It will default to a spherical one and if it is available it can use 
proj or gdal


On 8/6/15 3:09 PM, Anderson wrote:
>
> @scollis <https://github.com/scollis>, we are kind of going 
> back-wards, we should first solve issue #285 
> <https://github.com/ARM-DOE/pyart/issues/285>, then #294 
> <https://github.com/ARM-DOE/pyart/issues/294> and at last solve the 
> graphing. Of course, I was the one start with the last, so I can't 
> really complain about that.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ARM-DOE/pyart/pull/346#issuecomment-128491185>.
>

-- 
I ride for Parkinsons research http://www.events.org/sponsorship.aspx?id=51573

gamaanderson commented:
@jjhelmus  We could make a plot in aeqd, but you will not find maps in aeqd, at maximum coast- and country-lines. We could consider making a plot without map, something like `GridDisplay`, but that is an other issue.

jjhelmus commented:
I do not think we should use the aeqd projection to make a map, rather we should use tmerc to perform the calculation of the grid point locations in latitude and longitude.  Or in general use the same projection for transforming the grid points as is used in the map.  If these are not the same I would assume that there may be small shifts in the locations of grid points.  I'll do some tests and see if this is true.

gamaanderson commented:
if the grid as originally created with one projection, using **proj** in an other projection can create a error, a big one if the grid is big. It is for the grid creating function to inform what projection is being considered. And in the lack of a projection we should define which one is the standard (my vote go for aeqd), but any way this is also not for the `GridMapDisplay` to decide.

jjhelmus commented:
Right, I still am struggling with the idea that a grid must be created with a projections, but I'm sure it must.      With that then the two different projections make sense.  I need to do some more reading on the topic in order to better understand this.

gamaanderson commented:
Just added the alteration we were discussing:

* add projection options to `plot_basemap` and `_makea_basemap`
* remove pre-calculation of `self.grid_lons` e `self.grid_lats`
* add default basemap to `plot_crosshairs`, remove `_check_basemap` and the travis check for that

also I reconsidered and changed default from tranverse mercator to mercator, to match the traditional global maps, just need to set lat_ts to get the right scale.

jjhelmus commented:
The projection and kwargs are not being passed to the `_make_basemap` method.  **RadarMapDisplay** does a basic check on some of the parameters to see what arguments should be passed to Basemap (not all projects work with providing corner data), something like this would make sense here, see lines [221-237](https://github.com/ARM-DOE/pyart/blob/54d5f00d5c25db7f61edbcd23b4689b68556daec/pyart/graph/radarmapdisplay.py#L221-L237).  

I'm hesitant to change the default to mercator as the distortions at high latitudes are large.  One of ARM's majors sites is in Alaska and displaying the data on mercator projected maps is awful.  The default in the **RadarMapDisplay** class is Lambert conformal conic ('lcc').

gamaanderson commented:
The problem with Mercator is that it change the scales, but the lat_ts (latitude of true scale) correct this.

gamaanderson commented:
will Basemap not just ignore unneeded argument? it is necessary to test `using_corners`?

jjhelmus commented:
You are right on both points. I'm still learning about Basemap.  I think setting lat_ts to lat_0 makes sense. With that change defaulting to Mercator is fine with me.  Passing extra arguments with the default None value to Basemap also works, so the test for `using_corners` is not needed in the **RadarMapDisplay** class.

gamaanderson commented:
Does this last commit (2680650) cover this points? I not sure I understood everything correctly.

jjhelmus commented:
The Basemap defaults should be provide as non-None values for either width and height parameter or llcrnrlon, llcrnlat, urcrnrlon, urcrnrlat parameters but not both.  I cannot tell from the Basemap documentation which set of parameters would be used if both a supplied.  I would suggest calculating height and width if either is None and adding them to the Basemap kwargs is the corner parameters are not supplied.

gamaanderson commented:
Which parameter is used will depend on the projection, but for `merc` the corners are used (it warns that on execution)

jjhelmus commented:
This issue with having defaults for both sets is that a user would need to set all six to override the defaults.  I think most users would expect that setting one set would override the defaults of the other set.

gamaanderson commented:
done

jjhelmus commented:
Looks good to me!  Anything else needed before this is merged?

gamaanderson commented:
no, you can merge it.

jjhelmus commented:
Thanks for all the work @gamaanderson.  Glad to see this one merged.
